### ISSUE TITLE
UI Thread sometimes freezes on closing Windows using CompositingRenderer

### ISSUE THREAD

**Issue by HendrikMennen** [AvaloniaUI/Avalonia#9890]
Sometimes when I close a Popup or Window very quickly after opening it, the UI Thread freezes at this point, making the app unusable until I restart it. In my case this happens quite often when using the AvaloniaEdit CompletionWindow.
But I also had this happen on other Windows that use Animations (in my case a loading spinner), although its quite rare to happen.
Sadly I didn't find a way to reproduce it consistently yet.

11.0.0-preview3 net7
Windows 11
Maybe someone here knows more about it, @kekekeks ?

### Update
I usually had `UseWindowsUIComposition` turned off, to have the native rounded corners. When I leave it on default this isssue does not seem to happen.
Also it is Windows only, never happened on Linux/MacOS

![image](https://user-images.githubusercontent.com/25281882/210771126-98838dd2-64f8-4cc1-b137-49a137caf0fa.png)

**Comment by timunie** [comment]
Is `_compositor.Commit().Wait()` something that you call on your own?

**Comment by HendrikMennen** [comment]
No, its internal (inside CompositingRenderer) as you can see in the stack below.

**Comment by SimonCropp** [comment]
time to move to IAsyncDisposable?

**Comment by kekekeks** [comment]
Please, show stack traces from other threads.

**Comment by HendrikMennen** [comment]
I started using UseWindowsUIComposition=true now and this issue disappeared completely. 

> Please, show stack traces from other threads.


```
 @12712

 @20772

 @1556

 @12880

 @3916

 @20952

 @11744

.NET Long Running Task @14600
	Monitor.Wait()at C:\Users\hmenn\AppData\Roaming\JetBrains\Rider2022.3\resharper-host\SourcesCache\fac0d219dacf1ae7a75f907ae54bbdd67674f647e3bb82fe597028363a1bfa\Monitor.CoreCLR.cs:line 158
	Monitor.Wait()at C:\Users\hmenn\AppData\Roaming\JetBrains\Rider2022.3\resharper-host\SourcesCache\8f7f4140c11a35ef24bbf8273d3fc342a747466e6ce4c98cf1125ec9baa76\Monitor.cs:line 20
	ObserveOnObserverLongRunning<Unit>.Drain()
	ObserveOnObserverLongRunning<Unit>.<>c.<.cctor>b__17_0()
	TaskPoolScheduler.LongScheduledWorkItem<ObserveOnObserverLongRunning<Unit>>.<>c.<.ctor>b__3_0()
	ExecutionContext.RunInternal()
	Task.ExecuteWithThreadLocal()
	[Native to Managed Transition]

.NET ThreadPool Gate @17900
	WaitHandle.WaitOneNoCheck()
	PortableThreadPool.GateThread.GateThreadStart()
	[Native to Managed Transition]

.NET ThreadPool IO @10444
	[Managed to Native Transition]
	PortableThreadPool.IOCompletionPoller.Poll()
	[Native to Managed Transition]

.NET ThreadPool Worker @5228
	[Managed to Native Transition]
	Interop.Kernel32.GetQueuedCompletionStatus()
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	[Native to Managed Transition]

.NET ThreadPool Worker @21236
	[Managed to Native Transition]
	Interop.Kernel32.GetQueuedCompletionStatus()
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	[Native to Managed Transition]

.NET ThreadPool Worker @6744
	[Managed to Native Transition]
	[Native frame]
	Interop.Kernel32.ReadFile()
	RandomAccess.ReadAtOffset()
	SafeFileHandle.ThreadPoolValueTaskSource.ExecuteInternal()
	SafeFileHandle.ThreadPoolValueTaskSource.System.Threading.IThreadPoolWorkItem.Execute()
	ThreadPoolWorkQueue.Dispatch()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	[Native to Managed Transition]

.NET ThreadPool Worker @14888
	[Managed to Native Transition]
	Interop.Kernel32.GetQueuedCompletionStatus()
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	[Native to Managed Transition]

.NET ThreadPool Worker @13408
	[Managed to Native Transition]
	Interop.Kernel32.GetQueuedCompletionStatus()
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	[Native to Managed Transition]

.NET ThreadPool Worker @1956
	[Managed to Native Transition]
	Interop.Kernel32.GetQueuedCompletionStatus()
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	[Native to Managed Transition]

.NET ThreadPool Worker @12100
	[Managed to Native Transition]
	Interop.Kernel32.GetQueuedCompletionStatus()
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	[Native to Managed Transition]

.NET ThreadPool Worker @21444
	[Managed to Native Transition]
	Interop.Kernel32.GetQueuedCompletionStatus()
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	[Native to Managed Transition]

.NET ThreadPool Worker @17500
	[Managed to Native Transition]
	Interop.Kernel32.GetQueuedCompletionStatus()
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	[Native to Managed Transition]

.NET ThreadPool Worker @11732
	[Managed to Native Transition]
	Interop.Kernel32.GetQueuedCompletionStatus()
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	[Native to Managed Transition]

.NET ThreadPool Worker @1664
	[Managed to Native Transition]
	Interop.Kernel32.GetQueuedCompletionStatus()
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	[Native to Managed Transition]

.NET ThreadPool Worker @1760
	[Managed to Native Transition]
	Interop.Kernel32.GetQueuedCompletionStatus()
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	[Native to Managed Transition]

.NET ThreadPool Worker @11192
	[Managed to Native Transition]
	Interop.Kernel32.GetQueuedCompletionStatus()
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	[Native to Managed Transition]

.NET Timer @20116
	WaitHandle.WaitOneNoCheck()
	TimerQueue.TimerThread()
	[Native to Managed Transition]

Main Thread @7480
	Monitor.Wait()at C:\Users\hmenn\AppData\Roaming\JetBrains\Rider2022.3\resharper-host\SourcesCache\fac0d219dacf1ae7a75f907ae54bbdd67674f647e3bb82fe597028363a1bfa\Monitor.CoreCLR.cs:line 158
	ManualResetEventSlim.Wait()
	Task.SpinThenBlockingWait()
	Task.InternalWaitCore()
	Task.Wait()
	Task.Wait()
	CompositingRenderer.Dispose()at /_/src/Avalonia.Base/Rendering/Composition/CompositingRenderer.cs:line 297
	TopLevel.HandleClosed()
	WindowBase.HandleClosed()
	WindowImpl.AppWndProc() [2]
	WindowImpl.WndProc() [2]
	PopupImpl.WndProc()
	[Managed to Native Transition]
	WindowImpl.Dispose()
	PopupRoot.Dispose()
	Popup.<>c.<Open>b__91_8()
	AnonymousDisposable<(IPopupHost, CompositeDisposable)>.Dispose()
	Popup.PopupOpenState.Dispose()
	Popup.CloseCore()
	Popup.Close()
	CompletionWindowBase.Hide() {this: {CompletionWindow}}
	CompletionWindow.Collapse()
	CompletionWindow.CaretPositionChanged()
	Caret.OnDocumentUpdateFinished()
	TextArea.OnUpdateFinished()
	[Lightweight Method Call]
	MethodInvoker.Invoke()
	RuntimeMethodInfo.Invoke()
	Delegate.DynamicInvokeImpl()
	Delegate.DynamicInvoke()
	WeakEventManagerBase<TextDocumentWeakEventManager.UpdateFinished, TextDocument, EventHandler, EventArgs>.WeakHandlerList.DeliverEvent()
	WeakEventManagerBase<TextDocumentWeakEventManager.UpdateFinished, TextDocument, EventHandler, EventArgs>.PrivateDeliverEvent()
	WeakEventManagerBase<TextDocumentWeakEventManager.UpdateFinished, TextDocument, EventHandler, EventArgs>.DeliverEvent()
	TextDocument.EndUpdate()
	TextDocument.Replace()
	TextDocument.Insert()
	EmptySelection.ReplaceSelectionWithText()
	TextArea.ReplaceSelectionWithText()
	TextArea.PerformTextInput()
	TextArea.OnTextInput()
	InputElement.<>c.<.cctor>b__31_4()
	RoutedEvent<TextInputEventArgs>.<>c__DisplayClass1_0<InputElement>.<AddClassHandler>g__Adapter|0()
	RoutedEvent.<>c__DisplayClass23_0.<AddClassHandler>b__0()
	AnonymousObserver<(object, RoutedEventArgs)>.OnNextCore()
	ObserverBase<(object, RoutedEventArgs)>.OnNext()
	Subject<(object, RoutedEventArgs)>.OnNext()
	RoutedEvent.InvokeRaised()
	EventRoute.RaiseEventImpl()
	EventRoute.RaiseEvent()
	Interactive.RaiseEvent()
	KeyboardDevice.ProcessRawEvent()
	InputManager.ProcessInput()
	TopLevel.HandleInput()
	WindowImpl.AppWndProc() [1]
	WindowImpl.WndProc() [1]
	[Managed to Native Transition]
	Win32Platform.RunLoop()
	Dispatcher.MainLoop()
	ClassicDesktopStyleApplicationLifetime.Start()
	ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime<Avalonia.AppBuilder>()
	Program.Main()

PriorityScheduler:  @20708
	SemaphoreSlim.WaitUntilCountOrTimeout()
	SemaphoreSlim.Wait()
	BlockingCollection<Task>.TryTakeWithNoTimeValidation()
	BlockingCollection<Task>.<GetConsumingEnumerable>d__68.MoveNext()
	PriorityScheduler.<QueueTask>b__11_0()
	[Native to Managed Transition]

PriorityScheduler:  @9608
	SemaphoreSlim.WaitUntilCountOrTimeout()
	SemaphoreSlim.Wait()
	BlockingCollection<Task>.TryTakeWithNoTimeValidation()
	BlockingCollection<Task>.<GetConsumingEnumerable>d__68.MoveNext()
	PriorityScheduler.<QueueTask>b__11_0()
	[Native to Managed Transition]

PriorityScheduler:  @10472
	SemaphoreSlim.WaitUntilCountOrTimeout()
	SemaphoreSlim.Wait()
	BlockingCollection<Task>.TryTakeWithNoTimeValidation()
	BlockingCollection<Task>.<GetConsumingEnumerable>d__68.MoveNext()
	PriorityScheduler.<QueueTask>b__11_0()
	[Native to Managed Transition]

PriorityScheduler:  @1372
	SemaphoreSlim.WaitUntilCountOrTimeout()
	SemaphoreSlim.Wait()
	BlockingCollection<Task>.TryTakeWithNoTimeValidation()
	BlockingCollection<Task>.<GetConsumingEnumerable>d__68.MoveNext()
	PriorityScheduler.<QueueTask>b__11_0()
	[Native to Managed Transition]

PriorityScheduler:  @18768
	SemaphoreSlim.WaitUntilCountOrTimeout()
	SemaphoreSlim.Wait()
	BlockingCollection<Task>.TryTakeWithNoTimeValidation()
	BlockingCollection<Task>.<GetConsumingEnumerable>d__68.MoveNext()
	PriorityScheduler.<QueueTask>b__11_0()
	[Native to Managed Transition]

PriorityScheduler:  @6636
	SemaphoreSlim.WaitUntilCountOrTimeout()
	SemaphoreSlim.Wait()
	BlockingCollection<Task>.TryTakeWithNoTimeValidation()
	BlockingCollection<Task>.<GetConsumingEnumerable>d__68.MoveNext()
	PriorityScheduler.<QueueTask>b__11_0()
	[Native to Managed Transition]

PriorityScheduler:  @1876
	SemaphoreSlim.WaitUntilCountOrTimeout()
	SemaphoreSlim.Wait()
	BlockingCollection<Task>.TryTakeWithNoTimeValidation()
	BlockingCollection<Task>.<GetConsumingEnumerable>d__68.MoveNext()
	PriorityScheduler.<QueueTask>b__11_0()
	[Native to Managed Transition]

PriorityScheduler:  @12140
	SemaphoreSlim.WaitUntilCountOrTimeout()
	SemaphoreSlim.Wait()
	BlockingCollection<Task>.TryTakeWithNoTimeValidation()
	BlockingCollection<Task>.<GetConsumingEnumerable>d__68.MoveNext()
	PriorityScheduler.<QueueTask>b__11_0()
	[Native to Managed Transition]

PriorityScheduler:  @1824
	SemaphoreSlim.WaitUntilCountOrTimeout()
	SemaphoreSlim.Wait()
	BlockingCollection<Task>.TryTakeWithNoTimeValidation()
	BlockingCollection<Task>.<GetConsumingEnumerable>d__68.MoveNext()
	PriorityScheduler.<QueueTask>b__11_0()
	[Native to Managed Transition]

PriorityScheduler:  @4704
	SemaphoreSlim.WaitUntilCountOrTimeout()
	SemaphoreSlim.Wait()
	BlockingCollection<Task>.TryTakeWithNoTimeValidation()
	BlockingCollection<Task>.<GetConsumingEnumerable>d__68.MoveNext()
	PriorityScheduler.<QueueTask>b__11_0()
	[Native to Managed Transition]

PriorityScheduler:  @17788
	SemaphoreSlim.WaitUntilCountOrTimeout()
	SemaphoreSlim.Wait()
	BlockingCollection<Task>.TryTakeWithNoTimeValidation()
	BlockingCollection<Task>.<GetConsumingEnumerable>d__68.MoveNext()
	PriorityScheduler.<QueueTask>b__11_0()
	[Native to Managed Transition]

PriorityScheduler:  @6544
	SemaphoreSlim.WaitUntilCountOrTimeout()
	SemaphoreSlim.Wait()
	BlockingCollection<Task>.TryTakeWithNoTimeValidation()
	BlockingCollection<Task>.<GetConsumingEnumerable>d__68.MoveNext()
	PriorityScheduler.<QueueTask>b__11_0()
	[Native to Managed Transition]



```

**Comment by kekekeks** [comment]
I don't see anything that resembles a deadlock. It seems that composition batch somehow wasn't completed. Please check if there were any exceptions on the render thread (you can either enable "stop on all exceptions" in your ide of choice or enable logging)

**Comment by HendrikMennen** [comment]
Interesting, I never tried that.

While doing that I get this exception:
```
System.Runtime.InteropServices.COMException (0x80070057): CreateDesktopWindowTarget failed
   at Avalonia.Win32.WinRT.Impl.__MicroComICompositorDesktopInteropProxy.CreateDesktopWindowTarget(IntPtr hwndTarget, Int32 isTopmost) in /_/src/Windows/Avalonia.Win32/WinRT/WinRT.Generated.cs:line 5915
   at Avalonia.Win32.WinRT.Composition.WinUICompositorConnection.CreateWindow(IntPtr hWnd) in /_/src/Windows/Avalonia.Win32/WinRT/Composition/WinUICompositorConnection.cs:line 175
   at Avalonia.Win32.WinRT.Composition.WinUiCompositedWindowSurface.CreateGlRenderTarget() in /_/src/Windows/Avalonia.Win32/WinRT/Composition/WinUiCompositedWindowSurface.cs:line 34
   at Avalonia.Skia.GlRenderTarget..ctor(GRContext grContext, IGlPlatformSurface glSurface) in /_/src/Skia/Avalonia.Skia/Gpu/OpenGl/GlRenderTarget.cs:line 20
   at Avalonia.Skia.GlSkiaGpu.TryCreateRenderTarget(IEnumerable`1 surfaces) in /_/src/Skia/Avalonia.Skia/Gpu/OpenGl/GlSkiaGpu.cs:line 43
   at Avalonia.Skia.PlatformRenderInterface.CreateRenderTarget(IEnumerable`1 surfaces) in /_/src/Skia/Avalonia.Skia/PlatformRenderInterface.cs:line 215
   at Avalonia.Controls.TopLevel.CreateRenderTarget() in /_/src/Avalonia.Controls/TopLevel.cs:line 336
   at Avalonia.Controls.TopLevel.Avalonia.Rendering.IRenderRoot.CreateRenderTarget() in /_/src/Avalonia.Controls/TopLevel.cs:line 329
   at Avalonia.Rendering.Composition.Server.ServerCompositionTarget.Render() in /_/src/Avalonia.Base/Rendering/Composition/Server/ServerCompositionTarget.cs:line 88
   at Avalonia.Rendering.Composition.Server.ServerCompositor.RenderCore() in /_/src/Avalonia.Base/Rendering/Composition/Server/ServerCompositor.cs:line 122
   at Avalonia.Rendering.Composition.Server.ServerCompositor.Render() in /_/src/Avalonia.Base/Rendering/Composition/Server/ServerCompositor.cs:line 104
   at Avalonia.Rendering.RenderLoop.TimerTick(TimeSpan time) in /_/src/Avalonia.Base/Rendering/RenderLoop.cs:line 145
```

**Comment by HendrikMennen** [comment]
Interesting, didn't know of that.
```
Avalonia.OpenGL.OpenGlException: eglCreateWindowSurface failed with error EGL_BAD_NATIVE_WINDOW (0x300B)
   at Avalonia.OpenGL.Egl.EglPlatformOpenGlInterface.CreateWindowSurface(IntPtr window) in /_/src/Avalonia.OpenGL/Egl/EglPlatformOpenGlInterface.cs:line 56
   at Avalonia.OpenGL.Egl.EglGlPlatformSurface.CreateGlRenderTarget() in /_/src/Avalonia.OpenGL/Egl/EglGlPlatformSurface.cs:line 18
   at Avalonia.Skia.GlRenderTarget..ctor(GRContext grContext, IGlPlatformSurface glSurface) in /_/src/Skia/Avalonia.Skia/Gpu/OpenGl/GlRenderTarget.cs:line 20
   at Avalonia.Skia.GlSkiaGpu.TryCreateRenderTarget(IEnumerable`1 surfaces) in /_/src/Skia/Avalonia.Skia/Gpu/OpenGl/GlSkiaGpu.cs:line 43
   at Avalonia.Skia.PlatformRenderInterface.CreateRenderTarget(IEnumerable`1 surfaces) in /_/src/Skia/Avalonia.Skia/PlatformRenderInterface.cs:line 215
   at Avalonia.Controls.TopLevel.CreateRenderTarget() in /_/src/Avalonia.Controls/TopLevel.cs:line 336
   at Avalonia.Controls.TopLevel.Avalonia.Rendering.IRenderRoot.CreateRenderTarget() in /_/src/Avalonia.Controls/TopLevel.cs:line 329
   at Avalonia.Rendering.Composition.Server.ServerCompositionTarget.Render() in /_/src/Avalonia.Base/Rendering/Composition/Server/ServerCompositionTarget.cs:line 88
   at Avalonia.Rendering.Composition.Server.ServerCompositor.RenderCore() in /_/src/Avalonia.Base/Rendering/Composition/Server/ServerCompositor.cs:line 122
   at Avalonia.Rendering.Composition.Server.ServerCompositor.Render() in /_/src/Avalonia.Base/Rendering/Composition/Server/ServerCompositor.cs:line 104
   at Avalonia.Rendering.RenderLoop.TimerTick(TimeSpan time) in /_/src/Avalonia.Base/Rendering/RenderLoop.cs:line 145
```

**Comment by kekekeks** [comment]
Should be fixed by #10951
